### PR TITLE
Alter preferTags to use enum (Minimal container triples)

### PR DESCRIPTION
**Pull request by whikloj**
Setup new ITs.

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3334

# What does this Pull Request do?
Makes an new enum to allow for identifying include, exclude or silent prefer headers.

Made new `PrefersLdpIT` to hold prefer header tests. Moved several tests from `FedoraLdpIT`  to `PrefersLdpIT`

# How should this be tested?

As this is essentially about minimal container triples.
Make a resource, add some Rdf, and a contained resource.
GET the resource.
See user triples, server managed triples and containment triples.
GET the resource with `Prefer: return=representation; include="https://www.w3.org/ns/ldp#PreferMinimalContainer"`
See user triples and server managed triples.
GET the resource with `Prefer: return=representation; omit="https://www.w3.org/ns/ldp#PreferMinimalContainer"`
See containment triples.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers

### CODE REVIEW

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/rdf/LdpTriplePreferences.java [hunk 1]
```diff
@@ -25,37 +25,48 @@
 public interface LdpTriplePreferences {
 
     /**
-     * @return Whether to return a minimal container.
+     * What the prefer tag choice is.
      */
-    boolean getMinimal();
+    enum preferChoice {
```
Comment (awoods) [inline fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/rdf/LdpTriplePreferences.java:30]: This `enum` declaration can be simplified to:
```
    enum preferChoice {
        INCLUDE,
        EXCLUDE,
        SILENT
    }
```

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/services/EtagService.java [hunk 2]
```diff
@@ -74,21 +74,21 @@ public String getRdfResourceEtag(final String txId, final FedoraResource resourc
 
         // Factor in preferences which change which triples are included in the response
         etag.append('|');
-        if (prefers.prefersContainment()) {
+        if (!prefers.prefersContainment().equals(LdpTriplePreferences.preferChoice.EXCLUDE)) {
```
Comment (awoods) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/services/EtagService.java:77]: The double-negative probably warrants a clarifying, inline comment:
```
// Should containment be included, i.e. not excluded?
```

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/services/EtagService.java [hunk 3]
```diff
@@ -74,21 +74,21 @@ public String getRdfResourceEtag(final String txId, final FedoraResource resourc
...
             final var lastUpdated = containmentIndex.containmentLastUpdated(txId, resource.getFedoraId());
             if (lastUpdated != null) {
                 etag.append(lastUpdated);
             }
         }
         etag.append('|');
-        if (prefers.prefersMembership()) {
+        if (!prefers.prefersMembership().equals(LdpTriplePreferences.preferChoice.EXCLUDE)) {
```
Comment (awoods) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/services/EtagService.java:84]: // Should membership be included, i.e. not excluded?

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/PrefersLdpIT.java [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+        getReq.setHeader("Prefer", preferLink(
+                null,
+                PREFER_SERVER_MANAGED.getURI()));
+        try (final var dataset = getDataset(getReq)) {
+            final var graph = dataset.asDatasetGraph();
+            assertTrue(graph.contains(userRdfTriple));
+            assertFalse(graph.contains(systemManagedPropertyTriple));
+            // Containment is consider system managed.
+            assertFalse(graph.contains(containmentTriple));
+            assertTrue(graph.contains(membershipTriple));
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/PrefersLdpIT.java:291]: It seems that "membership" is just as "server-managed" as "containment", no?
..which argues for omitting "membership" triples here as well.

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/rdf/LdpTriplePreferences.java [hunk 5]
```diff
@@ -25,37 +25,48 @@
...
+            this.value = value;
+        }
+    }
 
     /**
-     * @return Whether this prefer tag demands membership triples.
+     * @return Whether this preference tag asks to include or omit a minimal container or is silent.
      */
-    boolean prefersMembership();
+    preferChoice preferMinimal();
```
Comment (awoods) [inline fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/rdf/LdpTriplePreferences.java:45]: The logic of the pre-PR interface is much simpler and readable than the `!preferences.preferX().equals(EXCLUDE)`, etc. Would it be possible to keep the existing interface and bury the new enum-based logic in the `LdpPreferTag` implementation?

**Review by awoods — CHANGES_REQUESTED** [review]

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/PrefersLdpIT.java (hunk 4 above)
Comment (whikloj) [reply fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/PrefersLdpIT.java:291]: I was going off of past tests. As there is a specific tag for including or excluding containment I think removing them them from the "system managed" category would be clearer. But 🤷

File: fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/domain/ldp/LdpPreferTag.java [hunk 6]
```diff
@@ -99,32 +101,58 @@ private preferChoice getChoice(final Property tag) {
...
+        // If we did ask for references. (Not shown by default).
+        return references.equals(INCLUDE);
     }
 
     @Override
-    public preferChoice prefersEmbed() {
-        return embed;
+    public boolean displayEmbed() {
+        // If we did ask for embedded resources. (Not shown by default).
+        return embed.equals(preferChoice.INCLUDE);
```
Comment (awoods) [inline fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/domain/ldp/LdpPreferTag.java:136]: May as well join the club, and drop the `preferChoice` qualifier.

**Review by awoods — CHANGES_REQUESTED** [review]

**Review by awoods — APPROVED** [review]

**Pull request merged**